packages/ngtools/webpack/src/angular_compiler_plugin.ts

```diff
@@ -593,6 +596,35 @@ export class AngularCompilerPlugin {
     }
   }
 
+  private _warnOnUnusedFiles(compilation: compilation.Compilation) {
+    // Only do the unused TS files checks when under Ivy
+    // since previously we did include unused files in the compilation
+    // See: https://github.com/angular/angular-cli/pull/15030
+    if (!this._compilerOptions.enableIvy) {
+      return;
+    }
+
+    const program = this._getTsProgram();
+    if (!program) {
+      return;
+    }
+
+    // Exclude the following files from unused checks
+    // - ngfactories & ngstyle might not have a corrispond
+    //   JS file example `@angular/core/core.ngfactory.ts`.
+    // - .d.ts files might not have a corrispond JS file due to bundling.
+    // - __ng_typecheck__.ts will never be requested.
+    const fileExcludeRegExp = /(\.(d|ngfactory|ngstyle)\.ts|ng_typecheck__\.ts)$/;
+
+    const unusedFilesWarning = program.getSourceFiles()
+      .filter(({ fileName }) => !fileExcludeRegExp.test(fileName) && !this._usedSourceFiles.has(fileName))
+      .map(({ fileName }) => `${fileName} is part of the TypeScript compilation but it's unused.`);
+
+    if (unusedFilesWarning.length) {
```

Comment:
Note: there is also the option to add these warnings in the loader, using `this.emitWarning` though I found that those messages are more verbose.

```
WARNING in ./app/app.component.ts
Module Warning (from C:/git/angular-cli/packages/ngtools/webpack/src/index.ts):
Module is part of the TypeScript compilation but it's unused.
```

@filipesilva, what do you think?